### PR TITLE
treewide: deprecate `libelf` package in favor of `elfutils`

### PR DESCRIPTION
**Pull request by philiptaron**
## Description of changes

Does almost everything for #271473.

> `libelf` did not see an update since 2014: https://github.com/Distrotech/libelf. `elfutils` is a popular API compatible ELF parser that most users expect.

As @Artturin notes: 
> Arch has the `libelf` package as an output of the `elfutils` package.
1. https://archlinux.org/packages/core/x86_64/libelf/
2. https://gitlab.archlinux.org/archlinux/packaging/packages/elfutils/-/blob/main/PKGBUILD?ref_type=heads#L142

## Things done

I made a commit per package changed. Some required more attention than others. A few required `zstd` added to their packages, as that's evidently a difference between `libelf` and `elfutils`.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by philiptaron** [comment]
I got everything except for pkgs/games/steam/fhsenv.nix, pkgs/os-specific/linux/kernel/manual-config.nix, and libelf itself.

File: pkgs/applications/misc/elf-dissector/default.nix [hunk 1]
```diff
@@ -2,23 +2,23 @@
...
 stdenv.mkDerivation rec {
   pname = "elf-dissector";
-  version = "unstable-2023-06-06";
+  version = "unstable-2023-11-25";
 
   src = fetchgit {
     url = "https://invent.kde.org/sdk/elf-dissector.git";
-    rev = "de2e80438176b4b513150805238f3333f660718c";
-    hash = "sha256-2yHPVPu6cncXhFCJvrSodcRFVAxj4vn+e99WhtiZniM=";
+    rev = "888ef6581df0400db45e0a7829a7336531b12641";
```
Comment (philiptaron) [inline pkgs/applications/misc/elf-dissector/default.nix:20]: upstream made fixes to enable this.

File: pkgs/applications/science/electronics/nvc/default.nix [hunk 2]
```diff
@@ -52,9 +48,9 @@ stdenv.mkDerivation rec {
   configureScript = "../configure";
 
   configureFlags = [
-    "--enable-vhpi"
     "--disable-lto"
   ];
+  dontDisableStatic = true;
```
Comment (philiptaron) [inline pkgs/applications/science/electronics/nvc/default.nix:53]: I added this to silence a build warning.

**Comment by Atemu** [comment]
Could you squash the tree-wide changes into one commit?

**Comment by philiptaron** [comment]
> Could you squash the tree-wide changes into one commit?

I can. I kept them as separate commits for bisection purposes.

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -25994,7 +26000,10 @@ with pkgs;
 
   # Clasp Common Lisp
   clasp-common-lisp = wrapLisp {
-    pkg = callPackage ../development/compilers/clasp { };
+    pkg = callPackage ../development/compilers/clasp {
+      stdenv = clang15Stdenv;
+      inherit (llvmPackages_15) llvm libclang;
```
Comment (philiptaron) [inline pkgs/top-level/all-packages.nix:26005]: [ofborg-eval](https://gist.github.com/GrahamcOfBorg/cb6cf2a15c58d427a368e1eed4c44f58) didn't like this. Hmm. I wonder why it works locally with `nix-build -A clasp-common-lisp`?

**Pull request closed without merging**